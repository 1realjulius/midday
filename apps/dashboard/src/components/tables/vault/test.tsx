"use client";

import { useVaultContext } from "@/store/vault/hook";
import { Button } from "@midday/ui/button";

export function Test() {
  const createFolder = useVaultContext((s) => s.createFolder);

  const handleCreateFolder = () => {
    createFolder({ name: "Untitled folder" });
  };

  return (
    <Button onClick={handleCreateFolder} variant="outline">
      Create folder
    </Button>
  );
}
