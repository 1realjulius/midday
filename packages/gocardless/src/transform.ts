import { capitalCase } from "change-case";

export const mapTransactionMethod = (method: string) => {
  switch (method) {
    case "Payment":
    case "Bankgiro payment":
    case "Incoming foreign payment":
      return "payment";
    case "Card purchase":
    case "Card foreign purchase":
      return "card_purchase";
    case "Card ATM":
      return "card_atm";
    case "Transfer":
      return "transfer";
    default:
      return "other";
  }
};

const transformName = (transaction) => {
  if (transaction?.additionalInformation) {
    return capitalCase(transaction.additionalInformation);
  }

  if (transaction?.remittanceInformationStructured) {
    return capitalCase(transaction.remittanceInformationStructured);
  }

  if (transaction?.remittanceInformationUnstructured) {
    return capitalCase(transaction.remittanceInformationUnstructured);
  }

  if (transaction?.creditorName) {
    return capitalCase(transaction.creditorName);
  }

  if (transaction?.debtorName) {
    return capitalCase(transaction?.debtorName);
  }

  if (transaction?.remittanceInformationUnstructuredArray?.at(0)) {
    return capitalCase(
      transaction.remittanceInformationUnstructuredArray?.at(0)
    );
  }

  console.log("No transaction name", transaction);
};

const transformDescription = (transaction, name) => {
  if (transaction?.remittanceInformationUnstructuredArray?.length) {
    const text = transaction?.remittanceInformationUnstructuredArray.join(" ");
    const description = capitalCase(text);

    // NOTE: Sometimes the description is the same as name
    // Let's skip that and just save if they are not the same
    if (description !== name) {
      return description;
    }
  }
};

export const transformTransactions = (transactions, { teamId, accountId }) => {
  // We want to insert transactions in reversed order so the incremental id in supabase is correct
  return transactions?.reverse().map((transaction) => {
    const method = mapTransactionMethod(
      transaction.proprietaryBankTransactionCode
    );

    let currencyExchange: { rate: number; currency: string } | undefined;

    if (Array.isArray(transaction.currencyExchange)) {
      const rate = Number.parseFloat(
        transaction.currencyExchange.at(0)?.exchangeRate ?? ""
      );

      if (rate) {
        const currency = transaction.currencyExchange.at(0)?.sourceCurrency;

        if (currency) {
          currencyExchange = {
            rate,
            currency,
          };
        }
      }
    }

    const name = transformName(transaction);

    return {
      date: transaction.valueDate,
      name,
      method: method || "unknown",
      internal_id: `${teamId}_${transaction.internalTransactionId}`,
      amount: transaction.transactionAmount.amount,
      currency: transaction.transactionAmount.currency,
      bank_account_id: accountId,
      category: transaction.transactionAmount.amount > 0 ? "income" : null,
      team_id: teamId,
      currency_rate: currencyExchange?.rate,
      currency_source: currencyExchange?.currency,
      balance: transaction?.balanceAfterTransaction?.balanceAmount?.amount,
      description: transformDescription(transaction, name),
      status: "posted",
    };
  });
};
