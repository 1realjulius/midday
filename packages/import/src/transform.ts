import crypto from "node:crypto";
import { capitalCase } from "change-case";
import type { Transaction } from "./types";
import { formatAmountValue, formatDate } from "./utils";

export function generateId(value: string) {
  const hash = crypto.createHash("sha256");
  hash.update(value);

  return hash.digest("hex");
}

export function transform(transaction: Transaction) {
  const internalId = generateId(
    `${transaction.date}-${transaction.description}`,
  );

  return {
    internal_id: `${transaction.teamId}_${internalId}`,
    team_id: transaction.teamId,
    status: "posted",
    method: "other",
    date: formatDate(transaction.date),
    amount: formatAmountValue(transaction.amount),
    name: transaction?.description && capitalCase(transaction.description),
    manual: true,
    category_slug: formatAmountValue(transaction.amount) > 0 ? "income" : null,
    bank_account_id: transaction.bankAccountId,
    currency: transaction.currency,
  };
}
