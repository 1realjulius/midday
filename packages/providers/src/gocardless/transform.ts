import { capitalCase } from "change-case";
import {
  Account as BaseAccount,
  Transaction as BaseTransaction,
} from "../types";
import {
  Transaction,
  TransactionDescriptionParams,
  TransformAccountName,
  TransformAccountParams,
  TransformTransactionParams,
} from "./types";

export const mapTransactionMethod = (method?: string) => {
  switch (method) {
    case "Payment":
    case "Bankgiro payment":
    case "Incoming foreign payment":
      return "payment";
    case "Card purchase":
    case "Card foreign purchase":
      return "card_purchase";
    case "Card ATM":
      return "card_atm";
    case "Transfer":
      return "transfer";
    default:
      return "other";
  }
};

export const transformTransactionName = (transaction: Transaction) => {
  if (transaction?.additionalInformation) {
    return capitalCase(transaction.additionalInformation);
  }

  if (transaction?.remittanceInformationStructured) {
    return capitalCase(transaction.remittanceInformationStructured);
  }

  if (transaction?.remittanceInformationUnstructured) {
    return capitalCase(transaction.remittanceInformationUnstructured);
  }

  if (transaction?.creditorName) {
    return capitalCase(transaction.creditorName);
  }

  if (transaction?.debtorName) {
    return capitalCase(transaction?.debtorName);
  }

  const remittanceInformation =
    transaction?.remittanceInformationUnstructuredArray?.at(0);

  if (remittanceInformation) {
    return capitalCase(remittanceInformation);
  }

  console.log("No transaction name", transaction);

  return "No information";
};

const transformDescription = ({
  transaction,
  name,
}: TransactionDescriptionParams) => {
  if (transaction?.remittanceInformationUnstructuredArray?.length) {
    const text = transaction?.remittanceInformationUnstructuredArray.join(" ");
    const description = capitalCase(text);

    // NOTE: Sometimes the description is the same as name
    // Let's skip that and just save if they are not the same
    if (description !== name) {
      return description;
    }
  }
};

export const transformTransaction = ({
  transaction,
  teamId,
  accountId,
}: TransformTransactionParams): BaseTransaction => {
  const method = mapTransactionMethod(
    transaction?.proprietaryBankTransactionCode
  );

  let currencyExchange: { rate: number; currency: string } | undefined;

  if (Array.isArray(transaction.currencyExchange)) {
    const rate = Number.parseFloat(
      transaction.currencyExchange.at(0)?.exchangeRate ?? ""
    );

    if (rate) {
      const currency = transaction?.currencyExchange?.at(0)?.sourceCurrency;

      if (currency) {
        currencyExchange = {
          rate,
          currency,
        };
      }
    }
  }

  const name = transformTransactionName(transaction);

  return {
    date: transaction.bookingDate,
    name,
    method,
    internal_id: `${teamId}_${transaction.internalTransactionId}`,
    amount: transaction.transactionAmount.amount,
    currency: transaction.transactionAmount.currency,
    bank_account_id: accountId,
    category: +transaction.transactionAmount.amount > 0 ? "income" : null,
    team_id: teamId,
    currency_rate: currencyExchange?.rate,
    currency_source: currencyExchange?.currency,
    balance: transaction?.balanceAfterTransaction?.balanceAmount?.amount,
    description: transformDescription({ transaction, name }),
    status: "posted",
  };
};

const transformAccountName = (account: TransformAccountName) => {
  if (account?.name) {
    return capitalCase(account.name);
  }

  if (account?.product) {
    return account.product;
  }

  if (account?.bank?.name) {
    return account.bank.name;
  }

  return "No name";
};

export const transformAccount = ({
  id,
  name,
  currency,
  bank,
  product,
}: TransformAccountParams): BaseAccount => {
  return {
    id,
    name: transformAccountName({ name, bank, product }),
    currency,
    institution: bank && {
      id: bank?.id,
      logo: bank?.logo,
      name: bank?.name,
    },
    provider: "gocardless",
  };
};
