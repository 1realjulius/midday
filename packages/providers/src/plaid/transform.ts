import { Account as BaseAccount } from "../types";
import { TransformAccountParams } from "./types";

export const transformAccount = ({
  id,
  name,
  currency,
  institution,
}: TransformAccountParams): BaseAccount => {
  return {
    id,
    name,
    currency,
    institution,
    provider: "plaid",
  };
};
