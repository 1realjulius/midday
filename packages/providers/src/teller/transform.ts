import { capitalCase } from "change-case";
import {
  Account as BaseAccount,
  Transaction as BaseTransaction,
} from "../types";
import { Transaction, TransformAccount, TransformTransaction } from "./types";

export const mapTransactionMethod = (type?: string) => {
  switch (type) {
    case "payment":
    case "bill_payment":
    case "digital_payment":
      return "payment";
    case "card_payment":
      return "card_purchase";
    case "atm":
      return "card_atm";
    case "transfer":
      return "transfer";
    case "ach":
      return "ach";
    case "interest":
      return "interest";
    case "deposit":
      return "deposit";
    case "wire":
      return "wire";
    case "fee":
      return "fee";
    default:
      return "other";
  }
};

export const mapTransactionCategory = (transaction: Transaction) => {
  if (transaction.type === "transfer") {
    return "transfer";
  }

  if (transaction.type === "fee") {
    return "fees";
  }

  // Positive values when money moves out of the account; negative values when money moves in.
  // For example, debit card purchases are positive; credit card payments, direct deposits, and refunds are negative.
  if (+transaction?.amount < 0) {
    return "income";
  }

  switch (transaction?.details.category) {
    case "bar":
    case "dining":
    case "groceries":
      return "meals";
    case "transport":
    case "transportation":
      return "travel";
    case "tax":
      return "taxes";
    case "office":
      return "office_supplies";
    case "phone":
      return "internet_and_telephone";
    case "software":
      return "software";
    case "entertainment":
    case "sport":
      return "activity";
    case "utilities":
    case "electronics":
      return "equipment";
    default:
      return null;
  }
};

const transformToSignedAmount = (amount: number) => {
  // Positive values when money moves out of the account; negative values when money moves in.
  // For example, debit card purchases are positive; credit card payments, direct deposits, and refunds are negative.
  if (amount > 0) {
    return -amount;
  }

  return amount * -1;
};

export const transformTransaction = ({
  transaction,
  teamId,
  bankAccountId,
}: TransformTransaction): BaseTransaction => {
  const method = mapTransactionMethod(transaction.type);

  return {
    date: transaction.date,
    name: transaction.description && capitalCase(transaction.description),
    description: null,
    method,
    internal_id: `${teamId}_${transaction.id}`,
    amount: transformToSignedAmount(+transaction.amount),
    currency: "USD",
    bank_account_id: bankAccountId,
    category: mapTransactionCategory(transaction),
    team_id: teamId,
    balance: transaction.running_balance,
    status: transaction?.status === "posted" ? "posted" : "pending",
  };
};

export const transformAccount = ({
  id,
  name,
  currency,
  institution,
  enrollment_id,
}: TransformAccount): BaseAccount => {
  return {
    id,
    name,
    currency,
    institution: {
      ...institution,
      logo: `https://teller.io/images/banks/${institution.id}.jpg`,
    },
    enrollment_id: enrollment_id,
    provider: "teller",
  };
};
