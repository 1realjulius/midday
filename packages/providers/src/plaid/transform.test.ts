import { expect, test } from "bun:test";
import { transformTransaction } from "./transform";

test("Transform pending transaction", () => {
  expect(
    transformTransaction({
      bankAccountId: "123",
      teamId: "123",
      transaction: {
        account_id: "AG7EkLW7DRSVaN8Z75jMT1DJN51QpWc9LKB7w",
        account_owner: null,
        amount: 5.4,
        authorized_date: "2024-02-23",
        authorized_datetime: null,
        category: ["Travel", "Taxi"],
        category_id: "22016000",
        check_number: null,
        counterparties: [
          {
            confidence_level: "VERY_HIGH",
            entity_id: "eyg8o776k0QmNgVpAmaQj4WgzW9Qzo6O51gdd",
            logo_url: "https://plaid-merchant-logos.plaid.com/uber_1060.png",
            name: "Uber",
            phone_number: null,
            type: "merchant",
            website: "uber.com",
          },
        ],
        date: "2024-02-24",
        datetime: null,
        iso_currency_code: "CAD",
        location: {
          address: null,
          city: null,
          country: null,
          lat: null,
          lon: null,
          postal_code: null,
          region: null,
          store_number: null,
        },
        logo_url: "https://plaid-merchant-logos.plaid.com/uber_1060.png",
        merchant_entity_id: "eyg8o776k0QmNgVpAmaQj4WgzW9Qzo6O51gdd",
        merchant_name: "Uber",
        name: "Uber 063015 SF**POOL**",
        payment_channel: "online",
        payment_meta: {
          by_order_of: null,
          payee: null,
          payer: null,
          payment_method: null,
          payment_processor: null,
          ppd_id: null,
          reason: null,
          reference_number: null,
        },
        pending: true,
        pending_transaction_id: null,
        personal_finance_category: {
          confidence_level: "VERY_HIGH",
          detailed: "TRANSPORTATION_TAXIS_AND_RIDE_SHARES",
          primary: "TRANSPORTATION",
        },
        personal_finance_category_icon_url:
          "https://plaid-category-icons.plaid.com/PFC_TRANSPORTATION.png",
        transaction_code: null,
        transaction_id: "NxkDjlyk45cQoDm5PEqJuKJaw6qrj9cy89zBA",
        transaction_type: "special",
        unofficial_currency_code: null,
        website: "uber.com",
      },
    })
  ).toMatchSnapshot();
});

// test("Transform card payment transaction", () => {
//   expect(
//     transformTransaction({
//       bankAccountId: "123",
//       teamId: "123",
//       transaction: {
//         type: "card_payment",
//         status: "posted",
//         running_balance: "83431.46",
//         links: {
//           self: "https://api.teller.io/accounts/acc_os41qe3a66ks2djhss000/transactions/txn_os41r5u90e29shubl2005",
//           account: "https://api.teller.io/accounts/acc_os41qe3a66ks2djhss000",
//         },
//         id: "txn_os41r5u90e29shubl2005",
//         details: {
//           processing_status: "complete",
//           counterparty: {
//             type: "organization",
//             name: "NORDSTROM",
//           },
//           category: "shopping",
//         },
//         description: "Nordstrom",
//         date: "2024-03-01",
//         amount: "-68.90",
//         account_id: "acc_os41qe3a66ks2djhss000",
//       },
//     })
//   ).toMatchSnapshot();
// });

// test("Transform income transaction", () => {
//   expect(
//     transformTransaction({
//       bankAccountId: "123",
//       teamId: "123",
//       transaction: {
//         type: "card_payment",
//         status: "posted",
//         running_balance: "83296.40",
//         links: {
//           self: "https://api.teller.io/accounts/acc_os41qe3a66ks2djhss000/transactions/txn_os41r5u90e29shubl2002",
//           account: "https://api.teller.io/accounts/acc_os41qe3a66ks2djhss000",
//         },
//         id: "txn_os41r5u90e29shubl2002",
//         details: {
//           processing_status: "complete",
//           counterparty: {
//             type: "organization",
//             name: "EXXON MOBIL",
//           },
//           category: "fuel",
//         },
//         description: "Exxon Mobil",
//         date: "2024-03-03",
//         amount: "-20.21",
//         account_id: "acc_os41qe3a66ks2djhss000",
//       },
//     })
//   ).toMatchSnapshot();
// });

// test("Transform type transfer", () => {
//   expect(
//     transformTransaction({
//       bankAccountId: "123",
//       teamId: "123",
//       transaction: {
//         type: "transfer",
//         status: "posted",
//         running_balance: "85897.25",
//         links: {
//           self: "https://api.teller.io/accounts/acc_os41qe3a66ks2djhss000/transactions/txn_os41r5ua0e29shubl2001",
//           account: "https://api.teller.io/accounts/acc_os41qe3a66ks2djhss000",
//         },
//         id: "txn_os41r5ua0e29shubl2001",
//         details: {
//           processing_status: "complete",
//           counterparty: {
//             type: "person",
//             name: "YOURSELF",
//           },
//           category: "general",
//         },
//         description: "Recurring Transfer to Savings",
//         date: "2024-01-27",
//         amount: "-37.99",
//         account_id: "acc_os41qe3a66ks2djhss000",
//       },
//     })
//   ).toMatchSnapshot();
// });
