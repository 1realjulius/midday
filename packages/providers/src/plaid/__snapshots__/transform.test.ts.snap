// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform pending transaction 1`] = `
{
  "amount": -5.4,
  "balance": null,
  "bank_account_id": "123",
  "category": "travel",
  "currency": "CAD",
  "date": "2024-02-24",
  "description": null,
  "internal_id": "123_NxkDjlyk45cQoDm5PEqJuKJaw6qrj9cy89zBA",
  "method": "other",
  "name": "Uber 063015 SF**POOL**",
  "status": "pending",
  "team_id": "123",
}
`;
